src/core/algorithms/fd/eulerfd/eulerfd.cpp

```diff
@@ -0,0 +1,427 @@
+        size_t inv_rhs_attr = inv_indexes[rhs_attr];
+        auto &tree = positive_cover_[inv_rhs_attr];
+        auto rhs = *schema_->GetColumn(rhs_attr);
+        tree.ForEach([&](Bitset const &lhs_attr) {
+            auto lhs = schema_->GetVertical(ChangeAttributesOrder(lhs_attr, attribute_indexes_));
+            RegisterFd(lhs, rhs);
+        });
+    }
+}
+
+void EulerFD::InitCovers() {
+    positive_cover_.clear();
+    positive_cover_.reserve(number_of_attributes_);
+    for (size_t i = 0; i < number_of_attributes_; i++) {
+        positive_cover_.push_back(SearchTreeEulerFD(number_of_attributes_));
+        positive_cover_.back().Add(Bitset(number_of_attributes_));
+    }
+
+    negative_cover_.clear();
+    negative_cover_.reserve(number_of_attributes_);
+    for (size_t i = 0; i < number_of_attributes_; i++) {
+        negative_cover_.push_back(SearchTreeEulerFD(number_of_attributes_));
+    }
+
+    attribute_frequences_.resize(number_of_attributes_);
+    std::fill(attribute_frequences_.begin(), attribute_frequences_.end(), 0);
+
+    attribute_indexes_.clear();
+}
+
+void EulerFD::BuildPartition() {
+    if (number_of_tuples_ == 0) {
+        return;
+    }
+    constant_columns_ = Bitset(number_of_attributes_);
+
+    // build partition: same values in each column unite into clusters,
+    // then stripping it: clusters with 1 element cant get us new invalid fd
+    for (size_t attr_num = 0; attr_num < number_of_attributes_; attr_num++) {
+        std::unordered_map<size_t, std::vector<size_t>> values;
+        for (size_t tuple_num = 0; tuple_num < number_of_tuples_; tuple_num++) {
+            auto value = tuples_[tuple_num][attr_num];
+            auto &similar_values = values[value];
+            similar_values.push_back(tuple_num);
+        }
+        if (values.size() == 1) {
+            // its mean in this column only one value
+            constant_columns_[attr_num] = true;
+            continue;
+        } else if (values.size() == number_of_tuples_) {
+            // its mean all clusters' sizes are 1, so we can pass cycle below
+            continue;
+        } else {
+            for (auto &&[_, cluster] : values) {
+                if (cluster.size() > 1) {
+                    clusters_.emplace_back(std::move(cluster), rand_function_);
+                }
+            }
+        }
+    }
+}
+
+EulerFD::Bitset EulerFD::BuildAgreeSet(size_t t1, size_t t2) const {
+    Bitset equal_attr(number_of_attributes_);
+    for (size_t attr_num = 0; attr_num < number_of_attributes_; ++attr_num) {
+        if (tuples_[t1][attr_num] == tuples_[t2][attr_num]) {
+            equal_attr.set(attr_num);
+        }
+    }
+    return equal_attr;
+}
+
+double EulerFD::SamlingInCluster(Cluster *cluster) {
+    return cluster->Sample([this](size_t t1, size_t t2) -> size_t {
+        Bitset agree_set = BuildAgreeSet(t1, t2);
+        auto &&[_, result] = invalids_.insert(agree_set);
```

Comment:
```suggestion
        auto &[_, result] = invalids_.insert(agree_set);
```

Reply:
Suggested changes won't compile, because `insert` return iterator and bool. Works `const auto &[_, result]`, but I don't see reason for use const lvalue reference instead rvalue reference in this case.